onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp

```diff
@@ -0,0 +1,502 @@
+            });
+
+            // c += a * b
+            UnrolledLoop<4>([&](size_t j) {
+                UnrolledLoop<NCols>([&](size_t i) { acc[i] = vfmaq_f32(acc[i], av[j], bv[i][j]); });
+            });
+
+            // increment b data pointers to next `SubBlkLen` elements
+            UnrolledLoop<NCols>([&](size_t i) { b_data[i] += 8; });
+        }
+
+        // increment pointers to next block
+        QuantBData += MlasQNBitBlkDataSizeInBytes(BlkBitWidth, BlkLen);
+        QuantBScale += 1;
+        QuantBZeroPointIdx += 1;
+    }
+
+    if constexpr (NCols == 4) {
+        float32x4_t sum = FoldAccumulators(acc[0], acc[1], acc[2], acc[3]);
+
+        if (BiasPtr != nullptr) {
+            sum = vaddq_f32(sum, vld1q_f32(BiasPtr));
+        }
+
+        vst1q_f32(SumPtr, sum);
+    } else {
+        for (size_t i = 0; i < NCols; ++i) {
+            SumPtr[i] = vaddvq_f32(acc[i]);
+            if (BiasPtr != nullptr) {
+                SumPtr[i] += BiasPtr[i];
+            }
+        }
+    }
+}
+
+}  // namespace
+
+//
+// MlasSQNBitGemmKernel and helpers.
+//
+
+template <size_t BlkBitWidth, size_t BlkLen>
+MLAS_FORCEINLINE size_t
+MlasSQNBitGemmKernelNeon(
+    const float* A,
+    const uint8_t* QuantBData,
+    const float* QuantBScale,
+    const uint8_t* QuantBZeroPoint,
+    float* C,
+    size_t CountM,
+    size_t CountN,
+    size_t CountK,
+    size_t lda,
+    size_t BlockStrideQuantB,
+    size_t ldc,
+    const float* Bias
+)
+{
+    [[maybe_unused]] auto impl0_reference = [&]() {
+        static_assert(BlkBitWidth == 4);
+
+        for (size_t m = 0; m < CountM; ++m) {
+            for (size_t n = 0; n < CountN; ++n) {
+                for (size_t k = 0, k_blk_idx = 0; k < CountK; k += BlkLen, k_blk_idx += 1) {
+                    const size_t kblocklen = std::min(CountK - k, BlkLen);
+
+                    const float b_s = QuantBScale[n * BlockStrideQuantB + k_blk_idx];
+                    const uint8_t b_z = [&]() -> uint8_t {
+                        if (QuantBZeroPoint != nullptr) {
+                            const size_t i = n * BlockStrideQuantB + k_blk_idx;
+                            uint8_t zp_packed = QuantBZeroPoint[i / 2];
+                            return ((i & 1) == 1) ? (zp_packed >> 4) : (zp_packed & 0x0F);
+                        } else {
+                            return 8;
+                        }
+                    }();
+                    const uint8_t* b_data =
+                        QuantBData + (n * BlockStrideQuantB + k_blk_idx) * MlasQNBitBlkDataSizeInBytes(BlkBitWidth, BlkLen);
+
+                    for (size_t kk = 0; kk < kblocklen; ++kk) {
+                        uint8_t b_packed = b_data[kk / 2];
+                        uint8_t b_byte = ((kk & 1) == 1) ? (b_packed >> 4) : (b_packed & 0x0F);
+                        float b_value = (b_byte - b_z) * b_s;
+
+                        C[m * ldc + n] += A[m * lda + k + kk] * b_value;
```

Comment:
>C[m * ldc + n] [](http://example.com/codeflow?start=24&length=14)

where is the reset of C? You need to set C to 0.f before the accumulation. We can not assume C is already reset

Reply:
added before k for loop